File: c/bj/bj3015.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>

#define GREATEST_POSSIBLE_NUM 500000
#define EMPTY -1

struct group_by_height {
	int height;
	int number;
};

int n;
int line[GREATEST_POSSIBLE_NUM];
struct group_by_height stack[GREATEST_POSSIBLE_NUM];
int top = EMPTY;

long count_pairs() {
	long cnt = 0;
	struct group_by_height new_group;
	for (int i = 0; i < n; ++i) {
		new_group.height = line[i];
		new_group.number = 1;
		while (top != EMPTY && stack[top].height <= line[i]) {
			cnt += stack[top].number;
			if (stack[top].height == line[i])
				new_group.number += stack[top].number;
			top--;
		}
		if (top != EMPTY)
			cnt++;
		stack[++top] = new_group;
	}
	return cnt;
}

int main() {
	scanf("%d", &n);
	for (int i = 0; i < n; ++i)
		scanf("%d", &line[i]);
	printf("%ld\n", count_pairs());
	return 0;
}

File: java/src/baekjoon/BJ3015.java
package baekjoon;

import java.io.*;
import java.util.Stack;

public class BJ3015 {

    private static class GroupByHeight {
        int height;
        int number;
        public GroupByHeight(int height, int number) {
            this.height = height;
            this.number = number;
        }
    }

    private static int n;
    private static int[] line;

    private static long countPairs() {
        Stack<GroupByHeight> stack = new Stack<>();
        long cnt = 0;

        GroupByHeight newGroup;
        for (int i = 0; i < n; ++i) {
            newGroup = new GroupByHeight(line[i], 1);
            while (!stack.isEmpty() && stack.peek().height <= line[i]) {
                cnt += stack.peek().number;
                if (stack.peek().height == line[i])
                    newGroup.number += stack.peek().number;

                stack.pop();
            }

            if (!stack.isEmpty())
                cnt++;

            stack.push(newGroup);
        }

        return cnt;
    }

    private static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        n = Integer.parseInt(br.readLine());
        line = new int[n];
        for (int i = 0; i < n; ++i)
            line[i] = Integer.parseInt(br.readLine());
        bw.write(String.valueOf(countPairs()));
        bw.flush();
    }

    public static void main(String[] args) throws IOException {
        solve();
    }

}
